File: packages/sdk/src/authentication.ts
import { QelosSDKOptions } from './types';
import BaseSDK from './base-sdk';

export interface IUser {
  _id: string;
  username: string;
  email: string;
  phone?: string;
  fullName: string;
  firstName: string;
  lastName: string;
  birthDate: string;
  roles: string[];
  metadata: any;

  [key: string]: any;
}

export interface ICredentials {
  username: string;
  password: string;
  roles?: string[];
}

export interface BasicPayload {
  user: IUser;
}

export interface ISignupInformation {
  username: string;
  email?: string;
  phone?: string;
  password: string;
  fullName?: string;
  firstName: string;
  lastName: string;
  birthDate: string;
}

export interface IApiToken {
  _id: string;
  nickname: string;
  tokenPrefix: string;
  expiresAt: string;
  lastUsedAt?: string;
  /** Workspace scope when set; may be an id string or a populated `{ _id, name }` from list responses */
  workspace?: string | { _id: string; name?: string } | null;
  /** Successful API-token authentications (server-maintained). */
  usageCount?: number;
  created: string;
}

export type SocialProvider = 'linkedin' | 'facebook' | 'google' | 'github';

export interface SocialLoginOptions {
  state?: string;
  returnUrl?: string;
}

export interface SocialAuthCallbackPayload {
  payload: BasicPayload & { workspace?: any };
  headers: { 'set-cookie': string | null };
}

export default class QlAuthentication extends BaseSDK {

  #refreshToken: string;
  #accessToken: string;
  #apiToken: string;

  get accessToken() {
    return this.#accessToken;
  }

  get isApiTokenAuth(): boolean {
    return !!this.#apiToken;
  }

  constructor(options: QelosSDKOptions) {
    super(options);
    if (options.accessToken || options.getAccessToken) {
      this.#accessToken = options.accessToken || options.getAccessToken();
      delete options.accessToken;
    }
    if (options.refreshToken) {
      this.#refreshToken = options.refreshToken;
      delete options.refreshToken;
    }
    if (options.apiToken) {
      this.#apiToken = options.apiToken;
    }
  }

  async signin(credentials: ICredentials) {
    const res = await this.callApi('/api/signin', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(credentials)
    });
    if (!res.ok) {
      throw new Error('failed to login')
    }
    const body = (await res.json()) as { payload: BasicPayload }

    return {
      ...body,
      headers: {
        'set-cookie': res.headers?.get('set-cookie')
      }
    }
  }

  oAuthSignin(credentials: ICredentials) {
    return this.callJsonApi<{ payload: BasicPayload & { token: string, refreshToken: string } }>(
      '/api/signin',
      {
        method: 'post',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ ...credentials, authType: 'oauth' })
      })
      .then(data => {
        this.#accessToken = data.payload.token;
        this.#refreshToken = data.payload.refreshToken;
        return data;
      })
  }

  signup(information: ISignupInformation) {
    return this.callJsonApi<{ payload: BasicPayload }>(
      '/api/signup',
      {
        method: 'post',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(information)
      })
  }

  oAuthSignup(information: ISignupInformation) {
    return this.callJsonApi<{ payload: BasicPayload & { token: string, refreshToken: string } }>(
      '/api/signup',
      {
        method: 'post',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ ...information, authType: 'oauth' })
      })
      .then(data => {
        this.#accessToken = data.payload.token;
        this.#refreshToken = data.payload.refreshToken;
        return data;
      })
  }

  refreshToken(refreshToken?: string) {
    refreshToken = refreshToken || this.#refreshToken;
    if (!refreshToken) {
      throw new Error('existing refresh token not found');
    }
    return this.callJsonApi<{ payload: BasicPayload & { token: string, refreshToken: string } }>(
      '/api/token/refresh',
      {
        method: 'post',
        headers: {
          authorization: 'Bearer ' + refreshToken
        }
      })
      .then(data => {
        this.#accessToken = data.payload.token;
        this.#refreshToken = data.payload.refreshToken;
        return data;
      })
  }

  async refreshCookieToken(cookieToken?: string) {
    const headers: Record<string, string> = {};
    if (cookieToken) {
      headers.authorization = 'Bearer ' + cookieToken;
    }
    const res = await this.callApi('/api/cookie/refresh', {
      method: 'post',
      headers,
    });
    if (!res.ok) {
      throw new Error('failed to refresh cookie token');
    }
    const body = (await res.json()) as {
      payload: BasicPayload & {
        cookieToken: string;
        workspace?: { _id: string } | null;
      };
    };
    return {
      ...body,
      headers: {
        'set-cookie': res.headers?.get('set-cookie'),
      },
    };
  }

  async apiTokenSignin(apiToken: string): Promise<IUser> {
    this.#apiToken = apiToken;
    this.qlOptions.apiToken = apiToken;
    return this.callJsonApi<IUser>('/api/me');
  }

  listApiTokens(): Promise<IApiToken[]> {
    return this.callJsonApi<IApiToken[]>('/api/me/api-tokens');
  }

  createApiToken(data: { nickname: string; workspace?: string; expiresAt: string }): Promise<{ token: string; apiToken: IApiToken }> {
    return this.callJsonApi<{ token: string; apiToken: IApiToken }>('/api/me/api-tokens', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(data),
    });
  }

  deleteApiToken(tokenId: string): Promise<void> {
    return this.callJsonApi<void>(`/api/me/api-tokens/${tokenId}`, {
      method: 'delete',
    });
  }

  logout() {
    return this.callApi('/api/logout', { method: 'post' })
  }

  getSocialLoginUrl(provider: SocialProvider, options: SocialLoginOptions = {}): string {
    const params: Record<string, string> = {};
    if (options.state) params.state = options.state;
    if (options.returnUrl) params.returnUrl = options.returnUrl;
    const qs = Object.keys(params).length ? `?${new URLSearchParams(params)}` : '';
    return this.buildUrl(`/api/auth/${provider}${qs}`);
  }

  getSocialCallbackUrl(provider: SocialProvider): string {
    return this.buildUrl(`/api/auth/${provider}/callback`);
  }

  startSocialLogin(provider: SocialProvider, options?: SocialLoginOptions): void {
    const location = (globalThis as any).location;
    if (!location) {
      throw new Error('startSocialLogin requires a browser environment');
    }
    location.href = this.getSocialLoginUrl(provider, options);
  }

  async exchangeAuthCallback(refreshToken: string): Promise<SocialAuthCallbackPayload> {
    if (!refreshToken) {
      throw new Error('refresh token is required');
    }
    const res = await this.callApi(`/api/auth/callback?rt=${encodeURIComponent(refreshToken)}`, {
      method: 'post'
    });
    if (!res.ok) {
      throw new Error('failed to exchange refresh token');
    }
    const body = (await res.json()) as { payload: BasicPayload & { workspace?: any } };
    return {
      ...body,
      headers: {
        'set-cookie': res.headers?.get('set-cookie')
      }
    };
  }

  getLoggedInUser() {
    return this.callJsonApi<IUser>('/api/me')
  }

  updateLoggedInUser(changes: Partial<IUser & { password?: string }>): Promise<IUser> {
    return this.callJsonApi<IUser>('/api/me', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(changes)
    })
  }
}

File: packages/python-sdk/qelos_sdk/authentication.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Literal, Optional
from urllib.parse import quote, urlencode

from .base_sdk import BaseSDK
from .types import QelosSDKOptions

SocialProvider = Literal["linkedin", "facebook", "google", "github"]


class QlAuthentication(BaseSDK):
    """Authentication module for sign-in, sign-up, token management, and user operations."""

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)
        self._refresh_token: Optional[str] = None
        self._access_token: Optional[str] = None
        self._api_token: Optional[str] = None

        if options.access_token or options.get_access_token:
            self._access_token = options.access_token or (
                options.get_access_token() if options.get_access_token else None
            )
            options.access_token = None

        if options.refresh_token:
            self._refresh_token = options.refresh_token
            options.refresh_token = None

        if options.api_token:
            self._api_token = options.api_token

    @property
    def access_token(self) -> Optional[str]:
        """Get the current access token."""
        return self._access_token

    @property
    def is_api_token_auth(self) -> bool:
        """Check if using API token authentication."""
        return bool(self._api_token)

    async def signin(self, credentials: Dict[str, Any]) -> Dict[str, Any]:
        """Sign in with username and password.

        Args:
            credentials: Dict with ``username``, ``password``, and optional ``roles``.
        """
        response = await self.call_api(
            "/api/signin",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(credentials),
        )
        if not response.is_success:
            raise Exception("failed to login")
        body = response.json()
        return {
            **body,
            "headers": {"set-cookie": response.headers.get("set-cookie")},
        }

    async def oauth_signin(self, credentials: Dict[str, Any]) -> Dict[str, Any]:
        """OAuth sign in. Sets access and refresh tokens automatically."""
        data = await self.call_json_api(
            "/api/signin",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps({**credentials, "authType": "oauth"}),
        )
        self._access_token = data["payload"]["token"]
        self._refresh_token = data["payload"]["refreshToken"]
        return data

    async def signup(self, information: Dict[str, Any]) -> Dict[str, Any]:
        """Sign up a new user.

        Args:
            information: Dict with ``username``, ``password``, ``firstName``,
                ``lastName``, ``birthDate``, and optional ``email``, ``phone``, ``fullName``.
        """
        return await self.call_json_api(
            "/api/signup",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(information),
        )

    async def oauth_signup(self, information: Dict[str, Any]) -> Dict[str, Any]:
        """OAuth sign up. Sets access and refresh tokens automatically."""
        data = await self.call_json_api(
            "/api/signup",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps({**information, "authType": "oauth"}),
        )
        self._access_token = data["payload"]["token"]
        self._refresh_token = data["payload"]["refreshToken"]
        return data

    async def refresh_token(self, refresh_token: Optional[str] = None) -> Dict[str, Any]:
        """Refresh the access token using a refresh token."""
        token = refresh_token or self._refresh_token
        if not token:
            raise Exception("existing refresh token not found")
        data = await self.call_json_api(
            "/api/token/refresh",
            method="POST",
            headers={"authorization": "Bearer " + token},
        )
        self._access_token = data["payload"]["token"]
        self._refresh_token = data["payload"]["refreshToken"]
        return data

    async def api_token_signin(self, api_token: str) -> Dict[str, Any]:
        """Sign in with an API token. Returns the current user."""
        self._api_token = api_token
        self._ql_options.api_token = api_token
        return await self.call_json_api("/api/me")

    async def list_api_tokens(self) -> List[Dict[str, Any]]:
        """List all API tokens for the current user."""
        return await self.call_json_api("/api/me/api-tokens")

    async def create_api_token(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new API token.

        Args:
            data: Dict with ``nickname``, ``expiresAt``, and optional ``workspace``.

        Returns:
            Dict with ``token`` and ``apiToken``.
        """
        return await self.call_json_api(
            "/api/me/api-tokens",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def delete_api_token(self, token_id: str) -> None:
        """Delete an API token by ID."""
        await self.call_json_api(f"/api/me/api-tokens/{token_id}", method="DELETE")

    async def logout(self) -> Any:
        """Log out the current user."""
        return await self.call_api("/api/logout", method="POST")

    def get_social_login_url(
        self,
        provider: SocialProvider,
        *,
        state: Optional[str] = None,
        return_url: Optional[str] = None,
    ) -> str:
        """Build the URL that initiates a social-auth flow for ``provider``.

        Args:
            provider: One of ``linkedin``, ``facebook``, ``google``, ``github``.
            state: Optional OAuth state value forwarded to the provider.
            return_url: Optional URL the callback page should redirect to after success.
        """
        params: Dict[str, str] = {}
        if state:
            params["state"] = state
        if return_url:
            params["returnUrl"] = return_url
        qs = ("?" + urlencode(params)) if params else ""
        return f"{self._app_url}/api/auth/{provider}{qs}"

    def get_social_callback_url(self, provider: SocialProvider) -> str:
        """The auth-service callback URL for ``provider``."""
        return f"{self._app_url}/api/auth/{provider}/callback"

    async def exchange_auth_callback(self, refresh_token: str) -> Dict[str, Any]:
        """Exchange a provider-issued refresh token for a Qelos cookie session.

        Calls ``POST /api/auth/callback?rt=<refresh_token>``. Returns the
        decoded payload and the upstream ``set-cookie`` header so callers
        can forward it to the end user.
        """
        if not refresh_token:
            raise Exception("refresh token is required")
        response = await self.call_api(
            f"/api/auth/callback?rt={quote(refresh_token, safe='')}",
            method="POST",
        )
        if not response.is_success:
            raise Exception("failed to exchange refresh token")
        body = response.json()
        return {
            **body,
            "headers": {"set-cookie": response.headers.get("set-cookie")},
        }

    async def get_logged_in_user(self) -> Dict[str, Any]:
        """Get the currently logged-in user."""
        return await self.call_json_api("/api/me")

    async def update_logged_in_user(self, changes: Dict[str, Any]) -> Dict[str, Any]:
        """Update the currently logged-in user.

        Args:
            changes: Partial user fields to update, may include ``password``.
        """
        return await self.call_json_api(
            "/api/me",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(changes),
        )
